bullet/src/Base.hh

```diff
@@ -0,0 +1,257 @@
+///    GenerateIdentity.
+/// 3) Hold explicit copies of raw pointers that can be deallocated
+
+// TO-DO(): Consider using unique_ptrs instead of shared pointers
+// for Bullet internal objects
+
+// Note: For Bullet library it's important the order in which the elements
+// are destroyed. The current implementation relies on C++ destroying the
+// elements in the opposite order stated in the structure
+struct WorldInfo
+{
+  std::string name;
+  std::shared_ptr<btDefaultCollisionConfiguration> collisionConfiguration;
+  std::shared_ptr<btCollisionDispatcher> dispatcher;
+  std::shared_ptr<btBroadphaseInterface> broadphase;
+  std::shared_ptr<btConstraintSolver> solver;
+  std::shared_ptr<btDiscreteDynamicsWorld> world;
+};
+
+struct ModelInfo
+{
+  std::string name;
+  Identity world;
+  bool fixed;
+  math::Pose3d pose;
+  std::vector<std::size_t> links = {};
+};
+
+struct LinkInfo
+{
+  std::string name;
+  Identity model;
+  math::Pose3d pose;
+  math::Pose3d inertialPose;
+  double mass;
+  btVector3 inertia;
+  std::shared_ptr<btDefaultMotionState> motionState;
+  std::shared_ptr<btCompoundShape> collisionShape;
+  std::shared_ptr<btRigidBody> link;
+};
+
+struct CollisionInfo
+{
+  std::string name;
+  std::shared_ptr<btCollisionShape> shape;
+  Identity link;
+  Identity model;
+  math::Pose3d pose;
+  bool isMesh;
+};
+
+struct JointInfo
+{
+  std::string name;
+  // Base class for all the constraint objects,
+  std::shared_ptr<btTypedConstraint> joint;
+  std::size_t childLinkId;
+  std::size_t parentLinkId;
+  int constraintType;
+  ignition::math::Vector3d axis;
+};
+
+inline btMatrix3x3 convertMat(Eigen::Matrix3d mat)
+{
+  return btMatrix3x3(mat(0, 0), mat(0, 1), mat(0, 2),
+                     mat(1, 0), mat(1, 1), mat(1, 2),
+                     mat(2, 0), mat(2, 1), mat(2, 2));
+}
+
+inline btVector3 convertVec(Eigen::Vector3d vec)
+{
+  return btVector3(vec(0), vec(1), vec(2));
+}
+
+inline Eigen::Matrix3d convert(btMatrix3x3 mat)
+{
+  Eigen::Matrix3d val;
+  val << mat[0][0], mat[0][1], mat[0][2],
+         mat[1][0], mat[1][1], mat[1][2],
+         mat[2][0], mat[2][1], mat[2][2];
+  return val;
+}
+
+inline Eigen::Vector3d convert(btVector3 vec)
+{
+  Eigen::Vector3d val;
+  val << vec[0], vec[1], vec[2];
+  return val;
+}
+
+class Base : public Implements3d<FeatureList<Feature>>
+{
+  public: std::size_t entityCount = 0;
+
+  public: inline std::size_t GetNextEntity()
+  {
+    return entityCount++;
+  }
+
+  public: inline Identity InitiateEngine(std::size_t /*_engineID*/) override
+  {
+    const auto id = this->GetNextEntity();
+    assert(id == 0);
+
+    return this->GenerateIdentity(0);
+  }
+
+  public: inline std::size_t idToIndexInContainer(std::size_t _id) const
+  {
+    std::size_t index = 0;
+    auto it = this->childIdToParentId.find(_id);
+    if (it != this->childIdToParentId.end())
+    {
+      for (const auto &pair : this->childIdToParentId)
```

Comment:
This works because `childIdToParentId` is an ordered map and so this iteration proceeds in ascending order of the keys of that map. Can you add a comment to this variable's declaration so someone doesn't change it to `unordered_map` sometime down the road?

Reply:
Added the comment in 024bdb2